### ISSUE TITLE
Problem with inherited Entities

### ISSUE THREAD

**Issue by sven5** [thepirat000/Audit.NET#7]
Hi,

I'm trying to integrate the Audit.NET library with my existing code. All of my entities inherit from an abstract base class. The code is not working and an error is thrown at the step "Get the entity type from the model that maps to the CLR type" in EntityKeyHelper.cs. The code is trying to find a CLR type for my base class, which cannot work properly because it doesn't exist in objectItemCollection. I'm not sure if it is correct to get the base entity type (GetBaseEntityType) for the given entity.

Regards
Sven

**Comment by philwindsor** [comment]
Hi, Sven 

Apologies if I've misunderstood your issue but have you tried adding some code to ignore your base class? As far as I am aware there a several ways

1. Use the fluent API

 `Audit.EntityFramework.Configuration.Setup()
.ForContext<PeopleDynamix.EF6.PeopleDynamixContext>()
.Reset()
.UseOptOut()
.IgnoreAny(t => t.IsAbstract);`

2. Use the ignore attribute on your base class `[Audit.EntityFramework.AuditIgnore]`

Thanks
Phil

**Comment by thepirat000** [comment]
I'm not able to reproduce the problem.
Can you share an example code?

It looks like Phil is right and you'll need to ignore the abstract class.

**Comment by sven5** [comment]
Hi Phil and Frederico,

I'm using Entity Framework 6.
I tried phil's suggestion but it's not working. I will try to explain more clear.
Imagine I have an entity that derives from an abstract base class. I load one of the entities and modify one property and call ctx.SaveChanges().
Exception is thrown on line 71 of EntityKeyHelper.cs - the call to Single() is failing with message "Sequence contains no elements".

Thanks 
Sven

**Comment by sven5** [comment]
Hi Phil and Frederico,

it seems I misunderstood the usage. I forgot to use/set the DataProvider.

Sorry for any confusion.

Sven

**Comment by philwindsor** [comment]
Hi Sven,

I too have encountered this error recently but haven't had the time to investigate further. Can you let me know how you resolved it please?

Thanks
Phil

**Comment by sven5** [comment]
Hi Phil,

the error disappeared after decorating the base class with AuditIgnore attribute. But I'm not sure if the audit event is created successfully - I hope. 
Let me explain my findings:
I realized that I have to use SqlServer DataProvider to save the audit logs to the DB. At the moment I stopped here because I'm currently unsure if Audit.NET framework fits to my usage scenario.

Regards
Sven

**Comment by thepirat000** [comment]
I'm guessing you want to save the audits in the same database as the audited data, right?

**Comment by thepirat000** [comment]
Ok. I think I've found the bug.

The method `GetBaseEntityType` on `EntityKeyHelper` is used to determine the entity type. 
And is *wrongly* assuming that your POCO entities does not inherits from a base class (i.e. inherits from object).

This was needed for compatibility with EDMX mapping, since EF creates proxies of your POCO entities, that are dynamic classes that inherits from your POCO classes.

https://github.com/thepirat000/Audit.NET/blob/master/src/Audit.EntityFramework/EntityKeyHelper.cs#L94

I will try to fix this today

**Comment by thepirat000** [comment]
Fixed on version 4.5.8.

Please use the latest package.

`GetBaseEntityType` method changed to:
```c#
private Type GetBaseEntityType(Type type)
{
    return ObjectContext.GetObjectType(type);
}
```

**Comment by sven5** [comment]
Hi Frederico,

thanks for resolving this issue. I will try it out now.
And yes you're right I'd like to store the audit log entries in the same database as my entities.

Regards
Sven
